fork-off/src/main.rs

```diff
@@ -12,168 +19,336 @@ pub struct Config {
-                debug!(target: "fork", "hashed prefix: {}, next key: {:?}", &prefix, &first_key);
+fn combine_states(mut state: Storage, initial_state: Storage, pallets: Vec<String>) -> Storage {
+    let pallets_prefixes: Vec<(String, String)> = pallets
+        .iter()
+        .map(|pallet| {
+            let hash = format!("0x{}", prefix_as_hex(pallet));
+            (pallet.clone(), hash)
+        })
+        .collect();
+    let mut cnt_removed_per_pallet: HashMap<String, usize> = pallets_prefixes
+        .iter()
+        .map(|(pallet, _)| (pallet.clone(), 0))
+        .collect();
+    let mut cnt_added_per_pallet = cnt_removed_per_pallet.clone();
+    state.retain(|k, _v| {
+        match pallets_prefixes
+            .iter()
+            .find(|(_, prefix)| is_prefix_of(prefix, k))
+        {
+            Some((pallet, _)) => {
+                *cnt_removed_per_pallet.get_mut(pallet).unwrap() += 1;
+                false
             }
+            None => true,
+        }
+    });
+    for (k, v) in initial_state.iter() {
+        if let Some((pallet, _)) = pallets_prefixes
+            .iter()
+            .find(|(_, prefix)| is_prefix_of(prefix, k))
+        {
+            *cnt_added_per_pallet.get_mut(pallet).unwrap() += 1;
+            state.insert(k.clone(), v.clone());
+        }
+    }
+    for (pallet, prefix) in pallets_prefixes {
+        info!(
+            "For pallet {} (prefix {}) Replaced {} entries by {} entries from initial_spec",
+            pallet, prefix, cnt_removed_per_pallet[&pallet], cnt_added_per_pallet[&pallet]
+        );
+    }
+    state
+}
 
-            stream::iter(pairs)
-        })
-        .flatten()
-        .collect()
-        .await
+#[tokio::main]
+async fn main() -> anyhow::Result<()> {
+    let Config {
+        http_rpc_endpoint,
+        initial_spec_path,
+        snapshot_path,
+        combined_spec_path,
+        use_snapshot_file,
+        pallets,
+        num_workers,
+    } = Config::parse();
+    env_logger::Builder::from_env(Env::default().default_filter_or("info")).init();
+
+    info!(target: "fork",
+        "Running with config: \n\
+        \thttp_rpc_endpoint: {}\n\
+        \tinitial_spec_path: {}\n\
+        \tsnapshot_path: {}\n\
+        \tcombined_spec_path: {}\n\
+        \tuse_snapshot_file: {}\n\
+        \tpallets: {:?}",
+        &http_rpc_endpoint, &initial_spec_path, &snapshot_path, &combined_spec_path, &use_snapshot_file, &pallets
+    );
+
+    let mut initial_spec: Value = serde_json::from_str(
+        &fs::read_to_string(&initial_spec_path).expect("Could not read chainspec file"),
+    )?;
+
+    assert_ne!(
+        initial_spec["genesis"]["raw"],
+        Value::Null,
+        "The initial provided chainspec must be raw! Make sure you use --raw when generating it."
+    );
+    if !use_snapshot_file {
+        let fetcher = StateFetcher::new(http_rpc_endpoint);
+        let block = fetcher.get_most_recent_block().await;
```

Comment:
you fetch that hash bu tthan it's not used anywhere (past logging it?)

Reply:
Oh, this is only some leftover from debugging :/ Removed it and added logging for best block at a different place.